code/modules/integrated_electronics/core/printer.dm

```diff
@@ -0,0 +1,448 @@
+			return TRUE
+		to_chat(user, "<span class='notice'>You install \the [O] into  \the [src]. </span>")
+		upgraded = TRUE
+		interact(user)
+		return TRUE
+
+	if(istype(O,/obj/item/disk/integrated_circuit/upgrade/clone))
+		if(can_clone)
+			to_chat(user, "<span class='warning'>\The [src] already has this upgrade. </span>")
+			return TRUE
+		to_chat(user, "<span class='notice'>You install \the [O] into  \the [src]. </span>")
+		can_clone = TRUE
+		interact(user)
+		return TRUE
+
+	return ..()
+
+/obj/item/device/integrated_circuit_printer/attack_self(var/mob/user)
+	interact(user)
+
+/obj/item/device/integrated_circuit_printer/interact(mob/user)
+	var/window_height = 600
+	var/window_width = 500
+
+	if(isnull(current_category))
+		current_category = recipe_list[1]
+
+	var/HTML = "<center><h2>Integrated Circuit Printer</h2></center><br>"
+	HTML += "Metal: [metal/metal_per_sheet]/[max_metal/metal_per_sheet] sheets.<br>"
+	HTML += "Circuits available: [upgraded ? "Advanced":"Regular"]."
+	HTML += "Assembly Cloning: [can_clone ? "Available": "Unavailable"]."
+	HTML += "Crossed out circuits mean that the printer is not sufficiently upgraded to create that circuit.<br>"
+	HTML += "<hr>"
+	if(can_clone)
+		HTML += "Here you can load script for your assembly.<br>"
+		if(as_printing)
+			HTML += " {Load Program} "
+		else
+			HTML += " <A href='?src=\ref[src];print=load'>{Load Program}</a> "
+		if(program == "blank")
+			HTML += " {Check Program} "
+		else
+			HTML += " <A href='?src=\ref[src];print=check'>{Check Program}</a> "
+		if((program == "blank")|as_printing)
+			HTML += " {Print assembly} "
+		else
+			HTML += " <A href='?src=\ref[src];print=print'>{Print assembly}</a> "
+		if(as_printing)
+			HTML += "<br> printing in process. Please insert more metal. "
+		HTML += "<br><hr>"
+	HTML += "Categories:"
+	for(var/category in recipe_list)
+		if(category != current_category)
+			HTML += " <a href='?src=\ref[src];category=[category]'>\[[category]\]</a> "
+		else // Bold the button if it's already selected.
+			HTML += " <b>\[[category]\]</b> "
+	HTML += "<hr>"
+	HTML += "<center><h4>[current_category]</h4></center>"
+
+	var/list/current_list = recipe_list[current_category]
+	for(var/obj/O in current_list)
+		var/can_build = TRUE
+		if(istype(O, /obj/item/integrated_circuit))
+			var/obj/item/integrated_circuit/IC = O
+			if((IC.spawn_flags & IC_SPAWN_RESEARCH) && (!(IC.spawn_flags & IC_SPAWN_DEFAULT)) && !upgraded)
+				can_build = FALSE
+		if(can_build)
+			HTML += "<A href='?src=\ref[src];build=[O.type]'>\[[O.name]\]</A>: [O.desc]<br>"
+		else
+			HTML += "<s>\[[O.name]\]: [O.desc]</s><br>"
+
+	user << browse(jointext(HTML, null), "window=integrated_printer;size=[window_width]x[window_height];border=1;can_resize=1;can_close=1;can_minimize=1")
+
+/obj/item/device/integrated_circuit_printer/Topic(href, href_list)
+	var/turf/T1=get_turf(usr)
+	var/turf/T2=get_turf(src)
+	if(!T1.Adjacent(T2))
```

Comment:
I dont think you need this adjacent check like this aft all but this js a quick look.